### PR TITLE
ENH: Adds selective hyperparameter optimization

### PR DESCRIPTION
**Pull request by itellaetxe**
# Summary
Users can now specify which parameters of the selected model can be hyperoptimized for. This way, for models with many hyperparameters, the search space is reduced a lot, whereas previously we would train $n^g$ models ($n:$ number of model parameters, $g:$ Hyperparameter grid points). Also fixes double-dipping problems in model fitting.

## Key Changes
- Addition of selection of hyperparameters to optimize in the CLI and its interactive version.
- the `hyperopt` model now trains on a dataset split instead of CV due to `hyperopt-sklearn` library limitations. The testing set indices for each model (dependent on covariates and systems) are returned in CSV files under the `ageml/model_age` directory.
- `AgeML.fit_age` method now has 2 train cases instead of 3, making it more concise and efficient.
- Modified the help message of the `-ht` argument.
- Adapted tests of interactive CLI for the new hyperparameter selection.

### Other changes
- Minor typo corrections
- `rng` from the `ageml.datasets` module is used in the `ui` module to do a index random permutation, instead of using `np.random`, which is discouraged
- MAE $\rightarrow$ mae

### CODE REVIEW

File: src/ageml/commands.py [hunk 1]
```diff
@@ -164,12 +164,28 @@ def configure_args(self, args):
...
+            args.hyperparameter_tuning = int(convert(hyperparam_tuning[0]))
+
+        hyperparameter_params = {}
+        if len(hyperparam_tuning) > 1:
+            for item in hyperparam_tuning[1:]:
+                if item.count("=") != 1:
+                    err_msg = (
+                        "Hyperparameter tuning parameters must be in the format "
+                        "param1=value1_low,value1_high param2=value2_low, value2_high..."
+                    )
```
Comment (JGarciaCondado) [inline src/ageml/commands.py:181]: should also check that always two values given a low and a high. What happens if someone gives C=1,2,3? this should through an error. As the user should write C=1,3 and ht=3 to have 3 hyperparameter points 1,2,3. Also ht should be a minimum of 2.

Comment (JGarciaCondado) [reply src/ageml/commands.py:181]: What about hyperparameters like kernels? where you want to choose different kernels. This should not be affected by ht.

File: src/ageml/messages.py [hunk 2]
```diff
@@ -108,8 +108,8 @@
 
 hyperparameter_grid_description = (
     "Number of points for which the hyperparameter optimization Grid Search will train\n"
-    "a model. The parameter ranges are predefined. An integer is required.\n"
-    "(e.g.   -ht 100   /   --hyperparameter_tuning 100)"
+    "a model, and parameter ranges to sample from. An integer is required, followed \n"
+    "by the parameters to optimize. (e.g.   -ht 100  C=1,2,3  kernel=linear,rbf)"
```
Comment (JGarciaCondado) [inline src/ageml/messages.py:112]: Should this not be only C=1,2? A low and a high value? Also how do you deal with if you put 3 kerrnel types. Also 100 hyperparameter grid points is unrealistic. Too many may lead users too put too many.

File: src/ageml/modelling.py [hunk 3]
```diff
@@ -206,26 +209,22 @@ def set_hyperparameter_grid(self):
...
-                    param_grid[f"model__{hyperparam_name}"] = np.rint(np.linspace(bounds[0], bounds[1],
-                                                                      int(self.hyperparameter_tuning))).astype(int)
-                elif hyperparam_types[hyperparam_name] == 'float':
-                    param_grid[f"model__{hyperparam_name}"] = np.linspace(bounds[0],
-                                                                          bounds[1],
-                                                                          int(self.hyperparameter_tuning))
+            for hyperparam_name in list(self.hyperparameter_params.keys()):
+                low, high = self.hyperparameter_params[hyperparam_name]
+                if hyperparam_types[hyperparam_name] == "log":
+                    param_grid[f"model__{hyperparam_name}"] = np.logspace(low, high, int(self.hyperparameter_tuning))
```
Comment (JGarciaCondado) [inline src/ageml/modelling.py:223]: Where are you telling the susers what are logs, ints etc...? Like if I choose C=-3,3 how do i know if it is using log? Or should I have to write C=0.001, 100?

File: src/ageml/modelling.py [hunk 4]
```diff
@@ -367,105 +367,127 @@ def fit_age(self, X, y):
...
-                                                        self.hyperparameter_grid,
-                                                        cv=self.CV_split,
-                                                        scoring="neg_mean_absolute_error")
-            opt_pipeline.fit(X, y)
-            print(f"Hyperoptimization best parameters: {opt_pipeline.best_params_}")
-            # Set best parameters in pipeline
-            self.pipeline.set_params(**opt_pipeline.best_params_)
+        if self.model_type != "hyperopt":
+            # Optimize hyperparameters if required
+            if self.hyperparameter_tuning > 0:
```
Comment (JGarciaCondado) [inline src/ageml/modelling.py:372]: hyperparameter tunnning should be at least 2 points? will give an error if you have C=1,2 and only have one point. Will it choose C=1 or C=2?

File: src/ageml/ui.py [hunk 5]
```diff
@@ -816,29 +819,49 @@ def model_age(self, df, model, tag):
...
+        """Gets and sets the train and test indices to keep track of them when running
+        the model_age pipeline. Necessary to inform the user that when hyperopt was
+        selected as the model, no CV was used, and instead a train test split was done.
+        """
+        data = self.dfs["cn"][covar][system]
+        X, y, _ = feature_extractor(data)
+
+        if self.args.model_type == "hyperopt":
+            # Split the data in training and test sets
+            # TODO: Substitute 0.2 factor by the user CLI argument
```
Comment (JGarciaCondado) [inline src/ageml/ui.py:852]: Why is this left as a todo instead of actually implementing it?

File: src/ageml/ui.py [hunk 6]
```diff
@@ -1761,7 +1800,9 @@ def model_age_command(self):
         print("Hyperparameter tuning. Number of points in grid search: (Default: 0)")
```
Comment (JGarciaCondado) [inline src/ageml/ui.py:1800]: Shoudl the default if we are actually running hypeparameter tunning not be at least 2? Something maybe like 5? Is not it should be saying hyperparameter tunning.

File: src/ageml/ui.py [hunk 7]
```diff
@@ -816,29 +819,49 @@ def model_age(self, df, model, tag):
...
+        if self.args.model_type == "hyperopt":
+            # Split the data in training and test sets
+            # TODO: Substitute 0.2 factor by the user CLI argument
+            test_size = int(0.2 * y.shape[0])
+            indices = rng.permutation(X.shape[0])
+            train_indices = indices[:-test_size]
+            test_indices = indices[-test_size:]
+
+        else:
+            train_indices = list(range(X.shape[0]))
```
Comment (JGarciaCondado) [inline src/ageml/ui.py:859]: why do we need this train and test indices? I see that they are the whole data structure.

**Review by JGarciaCondado — COMMENTED** [review]
For the most part LGTM. Minor changes are required to ensure edge cases are well handeld and that users understand the useag of hyperparameter optimization.

File: src/ageml/ui.py (hunk 7 above)
Comment (itellaetxe) [reply src/ageml/ui.py:859]: The reason behind it is to avoid writing `if` flow control statements every time I need to split the dataset depending on the model selection ('hyperopt' or not). It is an efficiency thing

File: src/ageml/ui.py (hunk 6 above)
Comment (itellaetxe) [reply src/ageml/ui.py:1800]: True, missed this

File: src/ageml/ui.py (hunk 5 above)
Comment (itellaetxe) [reply src/ageml/ui.py:852]: time

File: src/ageml/messages.py (hunk 2 above)
Comment (itellaetxe) [reply src/ageml/messages.py:112]: True, changed the C=1,2,3 to C=1,3. Also changed 100 to 10.
Working on dealing with string hyperparams.

File: src/ageml/modelling.py (hunk 3 above)
Comment (itellaetxe) [reply src/ageml/modelling.py:223]: I will add it in the documentation. And a link to the documentation explaining this.

File: src/ageml/modelling.py (hunk 4 above)
Comment (itellaetxe) [reply src/ageml/modelling.py:372]: True

File: src/ageml/modelling.py [hunk 8]
```diff
@@ -367,105 +367,127 @@ def fit_age(self, X, y):
...
+
+                # CV Summary of the selected hyperparameters
+                print("\nSummary metrics over all CV splits:")
+                summary_train = self.summary_metrics(split_metrics_train)
+                print("Train: MAE %.2f ± %.2f, RMSE %.2f ± %.2f, R2 %.3f ± %.3f, p %.3f ± %.3f" % tuple(summary_train))
+                summary_test = self.summary_metrics(split_metrics_test)
+                print("Test: MAE %.2f ± %.2f, RMSE %.2f ± %.2f, R2 %.3f ± %.3f, p %.3f ± %.3f" % tuple(summary_test))
+
+            # Select the best pipeline based on the mean scores
+            best_hyperparam_index = np.argmin(mae_means_test)
```
[the remark this answers is not in the document]
Comment (itellaetxe) [reply src/ageml/modelling.py:434]: True, solving it

File: src/ageml/modelling.py [hunk 9]
```diff
@@ -367,105 +372,132 @@ def fit_age(self, X, y):
...
+
+                    # Predictions on train and test set
+                    y_pred_train = cv_pipeline.predict(X_train)
+                    y_pred_test = cv_pipeline.predict(X_test)
+
+                    # Metrics in Train and Test sets
+                    split_metrics_train.append(self.calculate_metrics(y_train, y_pred_train))
+                    split_metrics_test.append(self.calculate_metrics(y_test, y_pred_test))
+
+                    # Print metrics
```
Comment (JGarciaCondado) [inline src/ageml/modelling.py:417]: Why are you priniting for each hypeparemeter optimization the train and test folds? Should only be printed for final best pipeline. Store this information and print at the end.

File: src/ageml/modelling.py [hunk 10]
```diff
@@ -367,105 +372,132 @@ def fit_age(self, X, y):
...
+                mae_means_test.append(mean_score_test)
+
+                # If the mean MAE is better than the previous best, save the results
+                if mean_score_test < best_split_mae:
+                    best_split_mae = mean_score_test
+                    pred_age = copy.deepcopy(temp_pred_age)
+                    corrected_age = copy.deepcopy(temp_corr_age)
+
+                # CV Summary of the selected hyperparameters
+                print("\nSummary metrics over all CV splits:")
```
Comment (JGarciaCondado) [inline src/ageml/modelling.py:441]: I believe here you are printing the metrics from the last CV point and not the best CV point. This should be changed.

File: src/ageml/modelling.py [hunk 11]
```diff
@@ -367,105 +372,132 @@ def fit_age(self, X, y):
...
+                    # Save results of hold out
+                    temp_pred_age[test] = y_pred_test
+                    temp_corr_age[test] = y_pred_test_no_bias
+
+                # Compute the mean of scores over all CV splits
+                mean_score_test = np.mean(split_metrics_test, axis=0)[0]  # Mean of MAE
+                mae_means_test.append(mean_score_test)
+
+                # If the mean MAE is better than the previous best, save the results
+                if mean_score_test < best_split_mae:
```
Comment (JGarciaCondado) [inline src/ageml/modelling.py:435]: It should be specificed in documentation that we are optimising with respect to MAE, could be pearson's correlation for example. Also a dict should be saved for each grid point with the MAE obtained. Maybe useful in the future to see MAE distributions across grid points don't change much.

Comment (itellaetxe) [reply src/ageml/modelling.py:435]: Added to documentation.

File: src/ageml/modelling.py (hunk 9 above)
Comment (itellaetxe) [reply src/ageml/modelling.py:417]: True. Now we only print the metrics of the BEST CV split of the BEST hyperparameters

**Review by JGarciaCondado — APPROVED** [review]
LGTM seems ready to approve and merge.
